indra/newview/llpanelface.h

```diff
@@ -57,6 +57,18 @@ class LLRadioGroup;
 class PBRPickerAgentListener;
 class PBRPickerObjectListener;
 
+// Need default_delete Template for C++23
+namespace std {
+    template<> struct default_delete<PBRPickerAgentListener> {
+        constexpr default_delete() noexcept = default;
+        void operator()(PBRPickerAgentListener* ptr) const noexcept;
+    };
+    template<> struct default_delete<PBRPickerObjectListener> {
+        constexpr default_delete() noexcept = default;
+        void operator()(PBRPickerObjectListener* ptr) const noexcept;
+    };
+}
+
```

Comment:
This part.
Is there a better way?

Reply:
Sorry for intermittent attention; I'm on Christmas holidays.

If I understand the problem and your solution, it's that `PBRPickerAgentListener` and `PBRPickerObjectListener` are incomplete types, so the default `std::default_delete`, which tries to `delete` the referents of those `std::unique_ptr`s, can't be instantiated in a translation unit that only sees the header file.

Your specializations of `std::default_delete` work around that by putting their `operator()` methods in the .cpp file, so the referents of the `std::unique_ptr`s can be destroyed without publishing their classes in the .h file.

The solution that comes to my mind would be to ensure that `~LLPanelFace()` is in the .cpp file - but it already is. Then I recalled that any assignment to either of those `unique_ptr`s would need to engage `default_delete` - but I saw no such assignment in the header file.

Maybe the issue arises from the memberwise assignments in the default copy constructor, copy assignment, move constructor and move assignment methods? I don't see explicit mention of any of those in the declaration of `LLPanelFace`, so presumably the compiler generates the default ones.

Have you tried declaring those four methods as `= delete` (without your `std::default_delete` specializations)?

Reply:
I haven't tried it yet. I will.

Reply:
I tried to define "operator()" within LLPanelFace, but a compilation error occurred.
Also, it seems that neither default nor delete can be defined for operator.

Reply:
I'm sorry, I seem to have confused things.

What I meant to suggest was that you try adding to the `LLPanelFace` class declaration:
```
LLPanelFace(const LLPanelFace&) = delete;  // copy constructor
LLPanelFace(LLPanelFace&&) = delete;       // move constructor
LLPanelFace& operator=(const LLPanelFace&) = delete;  // copy assignment operator
LLPanelFace& operator=(LLPanelFace&&) = delete;       // move assignment operator
```
Without explicit declarations, the compiler will generate default inline copy and move constructors and assignment operators, which operate on individual members. `LLPanelFace` contains a couple of `std::unique_ptr` members. Assigning to a `std::unique_ptr` requires deleting its previous referent, if any. So it could be that the generated assignment operators are responsible for the original incomplete-type errors.

With those declarations in place, you may discover other errors from some code (inadvertently?) trying to copy or move an `LLPanelFace` instance. If those actions are erroneous, you can fix them. If they're legit, you might have to code an actual copy or move assignment operator, placing the method definition in the .cpp file.

Reply:
i wil try.

Reply:
Simply adding it to the original code did not fix the error.

Reply:
Thank you for testing that.

(And Happy New Year!)